### ISSUE TITLE
`pattern_type_mismatch` false positive when `&str` is matched against a `const _: &str`

### ISSUE THREAD

**Issue by H2CO3** [rust-lang/rust-clippy#7946]
Lint name: `pattern_type_mismatch`

I tried this code ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3a64e0312d5d6be65ffd24ecd00bb284)):

```rust
#![deny(clippy::pattern_type_mismatch)]

const FOO: &str = "foo";

fn foo(s: &str) -> i32 {
    match s {
        FOO => 1,
        _ => 0,
    }
}
```

I expected to see this happen: the code should pass the lint, the type of the pattern is exactly the same as the type of the discriminant expression.

Instead, this happened: I got the "type of pattern does not match the expression type" error message.

### Meta

**Clippy version:** 0.1.58 (2021-11-07 46b8e74)
**Rust version:** 1.56.1 stable (on the Playground)

**Comment by pzheng1025** [comment]
I have faced the same issue.

**Comment by surechen** [comment]
Hi, I'm interested in this issue and want to follow up on it.
@rustbot claim

**Comment by surechen** [comment]
Hi, this issue has been fixed by [commit](https://github.com/rust-lang/rust-clippy/commit/8c1c763c2d5897c66f52bd5e7c16f48ae66c894c).

**Comment by xFrednet** [comment]
Hey @surechen, would you mind adding a test for this example to ensure that this never breaks again? Then we can close this for good :upside_down_face:

**Comment by surechen** [comment]
> Hey @surechen, would you mind adding a test for this example to ensure that this never breaks again? Then we can close this for good 🙃

Sorry to reply so late, I'm happy to do that.
